device/receive.go

```diff
@@ -455,21 +457,19 @@ func (peer *Peer) RoutineSequentialReceiver(maxBatchSize int) {
 				continue
 			}
 
-			peer.SetEndpointFromPacket(elem.endpoint)
+			validTailPacket = i
 			if peer.ReceivedWithKeypair(elem.keypair) {
+				peer.SetEndpointFromPacket(elem.endpoint)
```

Comment:
Is this required? It'll be called again after the loop because `validTailPacket` is set

Reply:
L464 calls `peer.SendStagedPackets()` which may send packets to the peer. So, I've retained the original order of operation of updating the Peer's endpoint prior.

This call *should* only occur during the early phase of a session, i.e. handshake, so the cost should be minimal. I favored the simplicity of duplicating the `peer.SetEndpointFromPacket()` call over more conditions.

Reply:
Perhaps the `peer.SendStagedPackets()` could also be deferred until after the batch is done, but I guess if it just happens once early on, it's not going to make a major difference unless the `peer.ReceivedWithKeypair` branch can be eliminated entirely.

Anyway this seems reasonable as is.